### PR TITLE
[16739] Operative SecureDS-SecureDS 

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

This PR intends to add necessary fixes to have a successful Secure DIscovery Server - Secure DIscovery Server interaction. Punctual changes on PDP Clients are also considered.
The example scenario in which fixes have been tested has been:
SimpleSecurePublisher - SDS1 - SDS2 - SImpleSecureSubscriber

Update 18/01/2022:
- In `EDPServer::process_and_release_change`, I still leave the call to `remove_change_from_writer_history()` (from PDPServer) since PDPServer makes use of it when checking if the change comes from a participant.

- I applied the shortcut to accept a writer entity id on `get_builtin_reader_history_pair_by_entity()`.

- Tested case Pub-DS1-DS2-Sub With and Without security.

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
**N/A** Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- [] Fast DDS test suite has been run locally. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** Documentation builds and tests pass locally. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 

- **N/A** Applicable backports have been included in the description.

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@richiprosima Please test discovery-server

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.hpp [hunk 1]
```diff
@@ -104,6 +104,30 @@ class EDPServer : public fastrtps::rtps::EDPSimple
...
+    /**
+     * This method removes all changes from the correct data writer history (change->witerGUID.enitity)
+     * and releases the change from the correct history dependeing whether the change is from
+     * this participant or remote
+     * @param change Pointer to the CacheChange_t.
+     * @return True if successful.
+     */
+    bool process_and_release_change(
+            fastrtps::rtps::CacheChange_t* change,
+            const bool& is_remote_participant = false);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.hpp:129]: No need for a const reference here:

```suggestion
            bool is_remote_participant = false);
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 2]
```diff
@@ -435,6 +435,83 @@ bool EDPServer::processLocalReaderProxyData(
...
+            }
+        }
+    }
+
+    return ret_val;
+}
+
+bool EDPServer::process_and_release_change(
+        fastrtps::rtps::CacheChange_t* change,
+        const bool& is_remote_participant)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp:477]: I think we can rename the parameter to show its true purpose.

```suggestion
        bool release_from_reader)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Two nits apart from what we have talked offline

**Comment by Mario-DL** [comment]
@richiprosima Please test discovery-server

**Comment by Mario-DL** [comment]
Just a detail to note. On [PDPServer.cpp lines 1438, 1450](https://github.com/eProsima/Fast-DDS/blob/feature/secure-discovery-server-poc/main/src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp) we were relasing the change in the same `remove_change_from_writer_history()` (passaing the `true` final parameter in the call), and if it failed, we then force the release. In our version, we always remove the change without relasing and then, we always call to release.

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 3]
```diff
@@ -1399,7 +1403,9 @@ void PDPServer::process_changes_release_(
             }
             else
             {
-                if (!edp->process_and_release_change(ch))
+                bool ret = (discovery_db_.is_writer(ch) || discovery_db_.is_reader(ch));
+
+                if (!ret || (ret && !edp->process_and_release_change(ch)))
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:1408]: Operator `||` has short-circuit mechanisms, so no need to check `ret` again:
```suggestion
                if (!ret || !edp->process_and_release_change(ch, false))
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.hpp [hunk 4]
```diff
@@ -126,7 +127,7 @@ class EDPServer : public fastrtps::rtps::EDPSimple
      */
     bool process_and_release_change(
             fastrtps::rtps::CacheChange_t* change,
-            const bool& is_remote_participant = false);
+            bool release_from_reader = false);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.hpp:130]: I like to force the users of an API to be explicit:
```suggestion
            bool release_from_reader);
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 5]
```diff
@@ -1424,8 +1430,9 @@ void PDPServer::process_changes_release_(
             }
             else
             {
-                if ((discovery_db_.is_writer(ch) || discovery_db_.is_reader(ch)) &&
-                        !edp->process_and_release_change(ch, true))
+                bool ret = (discovery_db_.is_writer(ch) || discovery_db_.is_reader(ch));
+
+                if (!ret || (ret && !edp->process_and_release_change(ch, true)))
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:1435]:
```suggestion
                if (!ret || !edp->process_and_release_change(ch, true))
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Some small requests

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**